File: core/security/src/main/java/org/apache/isis/core/security/authentication/Authentication.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package org.apache.isis.core.security.authentication;

import java.io.Serializable;

import org.apache.isis.applib.services.iactnlayer.ExecutionContext;
import org.apache.isis.applib.services.iactn.Interaction;
import org.apache.isis.applib.services.user.UserMemento;

/**
 * An immutable, serializable value type, that holds details about a user's authentication.
 *
 * <p>
 *     This is really little more than a thin wrapper around {@link ExecutionContext},
 *     surfaces a number of the security-specific attributes of that field.
 * </p>
 *
 * @apiNote This is a framework internal class and so does not constitute a formal API.
 *
 * @since 2.0 {@index}
 */
public interface Authentication extends Serializable {

    /**
     * The name of the authenticated user; for display purposes only.
     */
    default String getUserName() {
        return getUser().getName();
    }

    /**
     * A unique code given to this session during authentication.
     *
     * <p>
     * This can be used to confirm that this session has been properly created
     * and the user has been authenticated. It should return an empty string (
     * <tt>""</tt>) if this is unauthenticated user
     */
    String getValidationCode();

    /**
     * The (programmatically) simulated (or actual) user, belonging to this session.
     *
     * @apiNote immutable, allows an {@link Interaction} to (logically) run with its
     * own simulated (or actual) user
     */
    default UserMemento getUser() {
        return getExecutionContext().getUser();
    }

    /**
     * The {@link ExecutionContext} (programmatically) simulated (or actual), belonging to this session.
     *
     * @apiNote immutable, allows an {@link Interaction} to (logically) run with its
     * own simulated (or actual) clock
     */
    ExecutionContext getExecutionContext();

    /**
     * To support external security mechanisms such as keycloak, where the validity of the session is defined by
     * headers in the request.
     */
    default Type getType() {
        return Type.DEFAULT;
    }

    public enum Type {
        DEFAULT,
        /**
         * Instructs the {@link org.apache.isis.core.security.authentication.manager.AuthenticationManager} to not cache this session in its internal map of
         * sessions by validation code, and therefore to ignore this aspect when considering if an
         * {@link Authentication} is
         * {@link org.apache.isis.core.security.authentication.manager.AuthenticationManager#isSessionValid(Authentication) valid} or not.
         */
        EXTERNAL
    }

    // -- WITHERS

    /**
     * Returns a copy with given {@code executionContext}.
     * @param executionContext
     */
    Authentication withExecutionContext(ExecutionContext executionContext);
}

File: viewers/webcomponents/sample-vue-petclinic/frontend/src/authentication.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *       https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {shallowRef} from 'vue';

export interface AuthenticationContext {
  readonly username: string;
  readonly csrfHeaderName: string;
  readonly csrfParameterName: string;
  readonly csrfToken: string;
  readonly loginPath: string;
  readonly logoutPath: string;
}

interface GraphQLRequest {
  readonly document: string;
  readonly variables?: Record<string, unknown>;
  readonly operationName?: string;
  readonly signal?: AbortSignal;
}

export type GraphQLExecutor = (request: GraphQLRequest) => Promise<unknown>;

export const authenticationContext = shallowRef<AuthenticationContext | null>(null);
export const graphQlExecutor = shallowRef<GraphQLExecutor | undefined>(undefined);

export async function bootstrapAuthentication(endpoint = '/graphql'): Promise<void> {
  const contextEndpoint = document.querySelector<HTMLMetaElement>(
    'meta[name="causeway-authentication-context"]'
  )?.content.trim();
  if (!contextEndpoint) return;

  const response = await fetch(contextEndpoint, {
    credentials: 'same-origin',
    headers: {accept: 'application/json'},
    cache: 'no-store'
  });
  if (response.status === 401) {
    location.assign('/vue/login');
    return;
  }
  if (!response.ok) throw new Error(`Authentication context failed with HTTP ${response.status}.`);
  const context = await response.json() as AuthenticationContext;
  authenticationContext.value = Object.freeze(context);
  graphQlExecutor.value = createSecuredExecutor(endpoint, context);
}

function createSecuredExecutor(endpoint: string, context: AuthenticationContext): GraphQLExecutor {
  return async ({document, variables = {}, operationName, signal}: GraphQLRequest) => {
    const response = await fetch(endpoint, {
      method: 'POST',
      credentials: 'same-origin',
      headers: {
        accept: 'application/graphql-response+json, application/json',
        'content-type': 'application/json',
        [context.csrfHeaderName]: context.csrfToken
      },
      body: JSON.stringify({query: document, variables, operationName}),
      signal
    });
    if (response.status === 401) {
      const continuation = safeVueContinuation();
      const separator = context.loginPath.includes('?') ? '&' : '?';
      location.assign(context.loginPath + (continuation
        ? `${separator}continue=${encodeURIComponent(continuation)}`
        : ''));
      throw new Error('The authenticated session has expired.');
    }
    const text = await response.text();
    let body: unknown;
    try {
      body = text ? JSON.parse(text) : {};
    } catch (cause) {
      throw new Error(`GraphQL response was not valid JSON (HTTP ${response.status}).`, {cause});
    }
    if (!response.ok) {
      const error = new Error(`GraphQL request failed with HTTP ${response.status}.`) as Error & {status: number; responseBody: unknown};
      error.status = response.status;
      error.responseBody = body;
      throw error;
    }
    return body;
  };
}

function safeVueContinuation(): string | null {
  const candidate = `${location.pathname}${location.search}`;
  if (candidate === '/vue' || candidate === '/vue/' || candidate.startsWith('/vue/object/')) return candidate;
  return null;
}
